File: src/cas_auth.rs
fn after<'a>(s : &'a str, start: &'a str) -> Option<&'a str> {
    let pos = s.find(start)?;
    Some(&s[pos + start.len() ..])
}

fn before<'a>(s: &'a str, end: &'a str) -> Option<&'a str> {
    let pos = s.find(end)?;
    Some(&s[..pos])
}

fn between<'a>(s: &'a str, start: &'a str, end: &'a str) -> Option<&'a str> {
    after(s, start).and_then(|s| before(s, end))
}

fn parse_cas_response(body: &str) -> Option<&str> {
    if body.contains("<cas:authenticationSuccess>") {
        between(&body, "<cas:user>", "</")
    } else {
        None
    }
}

async fn http_get(url: &str) -> reqwest::Result<(reqwest::StatusCode, String)> {
    let res = reqwest::get(url).await?;
    Ok((res.status(), res.text().await?))
}

pub async fn validate_ticket(cas_prefix_url: &str, service: &str, ticket: &str) -> Result<String, String> {
    let url = format!("{}/serviceValidate?service={}&ticket={}", cas_prefix_url, service, ticket);
    println!("URL: {}", url);
    let (status, body) = http_get(&url).await.map_err(|err| err.to_string())?;
    if status == reqwest::StatusCode::OK {
        if let Some(user) = parse_cas_response(&body) {
            Ok(user.to_owned())
        } else {
            Err(body)
        }
    } else {       
        Err(format!("bad HTTP code {}", status))
    }
}

File: server/cas_auth.ts
// @ts-expect-error
import fetch from 'node-fetch';
import { throw_ } from './helpers';
import { Request, Response } from 'express';

const parse_cas_response = (body: string) => {
    const user = body.includes("<cas:authenticationSuccess>") && body.match(/<cas:user>(.*?)</)?.[1]
    if (!user) {
        console.error("CAS auth failed", body)
        throw_("CAS auth failed")
    }
    return user
}

const http_get = async (url: string) => {
    const res = await fetch(url);
    return { status: res.status, body: await res.text() }
}

export const validate_ticket = async (cas_prefix_url: string, service: string, ticket: string) => {
    const url = `${cas_prefix_url}/serviceValidate?service=${service}&ticket=${ticket}`
    console.log("URL:", url);
    const { status, body } = await http_get(url)
    if (status === 200) {
        return parse_cas_response(body)
    } else {       
        throw `bad HTTP code {status}`
    }
}

export const handle_single_logout = async (req: Request, res: Response) => {
    const m = req.body?.logoutRequest?.match(/<samlp:SessionIndex>(.*)<\/samlp:SessionIndex>/)
    if (m) {
        req.sessionStore.destroy(m[1]);
        res.status(204).end()
    } else {
        res.end()
    }
}
